peps/pep-0815.rst

```diff
@@ -0,0 +1,2360 @@
+``pip install jax`` (with no extra) leads to a broken installation for
+everybody and consequently dependency chains, a fundamental expected
+behavior in the Python ecosystem is dysfunctional.
+
+JAX includes 12 extra selectors to cover all use cases - many of which
+overlap and could be misleading to users if they don’t read in detail
+the documentation.
+
+It should be noted that most of these “extras” are technically mutually
+exclusive, though it is currently impossible to correctly express this
+incompatibility within the package metadata.
+
+.. code:: email
+
+    Provides-Extra: minimum-jaxlib
+    Provides-Extra: cpu
+    Provides-Extra: ci
+    Provides-Extra: tpu
+    Provides-Extra: cuda
+    Provides-Extra: cuda12
+    Provides-Extra: cuda13
+    Provides-Extra: cuda12-local
+    Provides-Extra: cuda13-local
+    Provides-Extra: rocm
+    Provides-Extra: k8s
+    Provides-Extra: xprof
+
+
+Bundled universal packages - monolithic builds
+''''''''''''''''''''''''''''''''''''''''''''''
+
+Including all possible variants in a single wheel is another option, but
+this leads to excessively large artifacts, wasting bandwidth and slower
+installation times for users who only need one specific variant. In some
+cases, such artifacts cannot be hosted on PyPI because they exceed its
+size limits.
+
+
+Wheel variant selection via source distribution
+'''''''''''''''''''''''''''''''''''''''''''''''
+
+`FlashAttention <https://github.com/Dao-AILab/flash-attention>`__ does
+not publish wheels on PyPI at all, but instead publishes a customized
+source distribution that performs platform detection, downloads the
+appropriate wheel from upstream server, and then provides it to the
+installer. Such an approach can provide a good end-user experience by
+selecting the most optimal variant automatically. However, it prevents
+``--only-binary`` installs from working and requires downloading source
+distribution and running its build phase. In this case, it also requires
+running with ``--no-build-isolation``. It requires hosting wheels
+separately, and does not provide for uniform experience across the
+ecosystem.
+
+**Induced Security Risk:** similarly to regular source builds, this
+model requires running arbitrary code at install time. The wheels
+are downloaded entirely outside package manager's control, extending
+the attack surface to two wheel fetching implementations and preventing
+proper provenance tracking.
+
+
+Ecosystem fragmentation
+'''''''''''''''''''''''
+
+The lack of standardized variant support has led to ecosystem
+fragmentation:
+
+**Inconsistent User Experience**: Each package uses different
+installation methods, creating confusion and reducing discoverability.
+
+**Development Tool Complications**: Installation tools, IDEs, and CI/CD
+systems struggle to handle non-standard installation requirements.
+
+**Documentation Burden**: Maintainers must create and maintain complex
```

Comment:
I would replace this with it being error prone. At least I haven't heard complaints about this being too much effort to document, but i've seen many cases in which those custom workarounds broke, from the pytorch index missing hashes over nvidia packages downloading the wrong artifact on the tooling side to so many complaints about installing GPU packages on the user side.